### PR TITLE
Add graph example with add node action

### PR DESCRIPTION
**Pull request by younik**
## Description

Add an example of sampling triangles starting from an empty graph, so requiring to add nodes. 
We never tested before the add node workflow, so there was a missing piece in mask computation.

- [x] I've read the [.github/CONTRIBUTING.md](.github/CONTRIBUTING.md) file
- [x] My code follows the typing guidelines
- [x] I've added appropriate tests
- [x] I've run pre-commit hooks locally

### CODE REVIEW

File: tutorials/examples/train_graph_triangle.py [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+            if g.edge_index.numel() > 0:
+                adj[g.edge_index[0], g.edge_index[1]] = 1
+                adj[g.edge_index[1], g.edge_index[0]] = 1
+            # Triangle has 3 undirected edges and degree 2 for each node
+            if (adj.sum() / 2 == 3) and torch.all(adj.sum(dim=1) == 2):
+                out[i] = self.reward_val
+        return out.view(*states.batch_shape)
+
+
+class GraphNodeEdgeAction(nn.Module):
```
Comment (younik) [inline tutorials/examples/train_graph_triangle.py:70]: We might move this in util & merge with the one we have

Comment (younik) [reply tutorials/examples/train_graph_triangle.py]: I integrated it in util, with a general `GraphNodeAction` module

File: src/gfn/gym/graph_building.py [hunk 2]
```diff
@@ -460,7 +460,7 @@ def __init__(
             edge_index=torch.zeros((2, 0), dtype=torch.long, device=device),
         )
         super().__init__(
-            num_node_classes=1,
+            num_node_classes=n_nodes,
```
[the remark this answers is not in the document]
Comment (younik) [reply src/gfn/gym/graph_building.py:463]: Yes, node class and node index are distinct concepts, and hopefully we will support node index soon for the backward policy, to remove node when we support multiple nodes of the same class

The semantics of this piece of code is correctly node_class, but we have the constraint that every node has a different class which makes it confusing. However, we can have `num_node_classes > n_nodes` (not in this specific environment as the number of nodes is fixed and I used arange for their class), for example `num_node_classes = 10` and `n_nodes = 3`, so it is not really an index.

File: src/gfn/states.py [hunk 3]
```diff
@@ -906,24 +910,23 @@ def forward_masks(self) -> TensorDict:
...
+        node_class_masks = node_class_masks.view(
+            *self.batch_shape, self.num_node_classes
+        )
         edge_masks = edge_masks.view(*self.batch_shape, max_possible_edges)
 
         # There are 3 action types: ADD_NODE, ADD_EDGE, EXIT
         action_type = torch.ones(
             *self.batch_shape, 3, dtype=torch.bool, device=self.device
         )
+        action_type[..., GraphActionType.ADD_NODE] = torch.any(node_class_masks, dim=-1)
```
[the remark this answers is not in the document]
Comment (younik) [reply src/gfn/states.py:924]: Here is node_class as well, not node index, as this is the mask for the policy head sampling the node class

File: src/gfn/states.py [hunk 4]
```diff
@@ -980,29 +987,23 @@ def backward_masks(self) -> TensorDict:
...
 
         # There are 3 action types: ADD_NODE, ADD_EDGE, EXIT
         action_type = torch.zeros(
             *self.batch_shape, 3, dtype=torch.bool, device=self.device
         )
-        action_type[..., GraphActionType.ADD_NODE] = torch.tensor(
-            [graph.x is not None and graph.x.size(0) > 0 for graph in self.data.flat],
-            device=self.device,
-        ).view(self.batch_shape)
+        action_type[..., GraphActionType.ADD_NODE] = torch.any(node_class_masks, dim=-1)
```
Comment (younik) [inline src/gfn/states.py:1001]: this will be called node index instead,  as the backward policy will sample index instead of class

File: src/gfn/utils/modules.py [hunk 5]
```diff
@@ -288,45 +289,24 @@ def __init__(
...
-        self._input_dim = 1  # Each node input is a single integer before embedding.
-        self._n_nodes = n_nodes
+        self.num_node_classes = num_node_classes
         self.hidden_dim = self.embedding_dim = embedding_dim
         self.is_backward = is_backward
         self.is_directed = directed
         self.num_conv_layers = num_conv_layers
         self.num_edge_classes = num_edge_classes
 
-        # Output dimension.
```
[the remark this answers is not in the document]
Comment (younik) [reply src/gfn/utils/modules.py:307]: This was to compute the output dimension, which is not needed (and was wrong as the output is a dict), and it is dynamic for this

File: src/gfn/utils/modules.py [hunk 6]
```diff
@@ -394,38 +396,37 @@ def __init__(
...
-        node_features, batch_ptr = (states_tensor.x, states_tensor.ptr)
+        node_features = states_tensor.x
+        B = len(states_tensor)
+        device = node_features.device
+
+        # Embed node classes
+        if node_features.numel() > 0:
+            x = self.embedding(node_features.squeeze(-1))
+        else:
+            x = torch.zeros(0, self.hidden_dim, device=device)
```
[the remark this answers is not in the document]
Comment (younik) [reply src/gfn/utils/modules.py:427]: This is to handle the case where the graph has no nodes; we use zeros as features, so we can continue the forward pass.

[the remark this answers is not in the document]
Comment (josephdviviano) [reply src/gfn/utils/modules.py:429]: I added a comment, thanks!

**Review by josephdviviano — APPROVED** [review]
Thank you @younik - and sorry for the delay on my review. Excellent work as usual, and I appreciate the speed.
